File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.js
class Node{
    constructor(v){
        this.left = null;
        this.value = v;
        this.right = null;
    }
}

function solution(nodeinfo) {
    const mappedInfo = nodeinfo.map((v,i)=>({id:i+1, x:v[0], y:v[1]}));
    const sortedInfo = mappedInfo.sort((a,b)=>b.y-a.y)
    
    const tree = new Node(sortedInfo[0])
    
    for(let i=1; i<sortedInfo.length; i+=1){
        const info = sortedInfo[i];
        let curValue = tree
        while(true){
            if(curValue.value.x<info.x){
                // 오른쪽으로
                if(curValue.right===null){
                    curValue.right = new Node(info)
                    break;
                }else{
                    curValue = curValue.right;
                }
            }
            else if(curValue.value.x>info.x){
                // 왼쪽으로
                if(curValue.left===null){
                    curValue.left = new Node(info)
                    break;
                }else{
                    curValue = curValue.left;                    
                }
            }
        }
    }
    
    function preorder(node){
        // 나, 왼, 오
        preorderResult.push(node.value.id)
        if(node.left){
            preorder(node.left)
        }
        if(node.right){
            preorder(node.right)
        }
    }
    function postorder(node){
        // 왼, 오, 나
        if(node.left){
            postorder(node.left)
        }
        if(node.right){
            postorder(node.right)
        }
        postOrderResult.push(node.value.id)
    }
    
    const preorderResult = [];
    const postOrderResult = []
    preorder(tree)
    postorder(tree)
    return [preorderResult, postOrderResult]
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.*;

class Solution {
    
    List<List<Integer>> answer = new ArrayList<>();

    public int[][] solution(int[][] nodeinfo) {
        int n = nodeinfo.length;
        answer.add(new ArrayList<>());
        answer.add(new ArrayList<>());
        int [][] info = new int[n][3];
        for(int i=0; i<n; i++) {
            int x = nodeinfo[i][0];
            int y = nodeinfo[i][1];
            int idx = i+1;
            info[i][0] = x;
            info[i][1] = y;
            info[i][2] = idx;
        }
        Arrays.sort(info, (a, b)->{
            if(b[1]==a[1]) {
                return Integer.compare(a[0], b[0]);
            }
            return Integer.compare(b[1], a[1]);
        });

        Node root = new Node(null, info[0][0], info[0][2], null);

        for (int i = 1; i < n; i++) {
            int x = info[i][0];
            int y = info[i][1];
            int idx = info[i][2];

            Node cur = root;
            Node newNode = new Node(null, x, idx, null);
            while (true) {
                if (cur.value > x) {
                    // 왼쪽
                    if (cur.left == null) {
                        cur.left = newNode;
                        break;
                    }
                    cur = cur.left;
                } else {
                    // 오른쪽
                    if(cur.right == null) {
                        cur.right = newNode;
                        break;
                    }
                    cur = cur.right;
                }
            }
        }
        preorder(root);
        postorder(root);
        int[][] result = new int[2][n];

        for (int i = 0; i < n; i++) {
            result[0][i] = answer.getFirst().get(i);
        }

        for (int i = 0; i < n; i++) {
            result[1][i] = answer.get(1).get(i);
        }
        return result;
    }

    public void preorder(Node node) {
        if (node == null) return;
        answer.getFirst().add(node.idx);
        preorder(node.left);
        preorder(node.right);
    }

    public void postorder(Node node) {
        if (node == null) return;
        postorder(node.left);
        postorder(node.right);
        answer.get(1).add(node.idx);
    }

    public static class Node {
        Node left;
        int value;
        int idx;
        Node right;

        public Node(Node left, int value, int idx, Node right) {
            this.left = left;
            this.value = value;
            this.idx = idx;
            this.right = right;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "left=" + left +
                    ", value=" + value +
                    ", idx=" + idx +
                    ", right=" + right +
                    '}';
        }
    }
}